### PR TITLE
Feat/48 instance db persistence

### LINKED ISSUE #48: Add persistence to instance creation
Add table persisiting the instances for each user, as well as he preset instances available

### PR DESCRIPTION
**Pull request by coffeemakingtoaster**
# Changes
- add persistence layer to instances
- add get endpoints for instances

Closes #48

### CODE REVIEW

File: backend/pkg/db/tables/instances/helper.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package db_instances
+
+func getInstanceId(usersub string, containerUUID string) string {
```
Comment (kontr0x) [inline backend/pkg/db/tables/instances/helper.go:3]: Be consistent with variable names. E.g. "usersub" in helper.go and helper_test.go

File: backend/pkg/db/tables/instances/instances.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+		TableName: aws.String(TABLE_NAME),
+		Item:      marshalledItem,
+	}
+
+	_, err := conn.PutItem(context.TODO(), &param)
+
+	if err != nil {
+		log.Warn().Msgf("Could not insert instance item %v", err)
+	} else {
+		log.Warn().Msg("Created")
```
Comment (kontr0x) [inline backend/pkg/db/tables/instances/instances.go:117]: Is warning log here necessary? Make sure to improve log message content.

File: backend/pkg/db/tables/instances/instances.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	}
+	return err
+}
+
+func GetAllUserInstances(usersub string) ([]InstancesTable, error) {
```
Comment (kontr0x) [inline backend/pkg/db/tables/instances/instances.go:122]: Ensure consistency of variable names.

File: backend/pkg/db/tables/instances/instances.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+		params.ExclusiveStartKey = scanResult.LastEvaluatedKey
+	}
+
+	return result, nil
+}
+
+func GetInstanceById(userSub string, containerUUID string) (InstancesTable, error) {
+
+	log.Warn().Msg("askdjlaskldhasdjashjdk")
```
Comment (kontr0x) [inline backend/pkg/db/tables/instances/instances.go:176]: Sorry I don't speak gibberish.

File: backend/pkg/instances/handler.go [hunk 5]
```diff
@@ -79,6 +107,26 @@ func CreatePresetContainerHandler(c *gin.Context) {
 		return
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:107]: Unnecessary return statement.

File: backend/pkg/instances/handler.go [hunk 6]
```diff
@@ -79,6 +107,26 @@ func CreatePresetContainerHandler(c *gin.Context) {
...
+			Tag:  "placeholder",
+		},
+		InstanceStatus: db_instances.STATUS_VALUE_SCHEDULED,
+		CreatedAt:      time.Now().Format(time.RFC3339),
+		StartedAt:      "N/A",
+	})
+
+	if err != nil {
+		util.ThrowInternalServerErrorException(c, "Could not save item at the current time")
+		return
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:127]: Unnecessary return statement.

File: backend/pkg/instances/handler.go [hunk 7]
```diff
@@ -133,6 +181,27 @@ func CreateCustomContainerHandler(c *gin.Context) {
...
+		},
+		InstanceStatus: db_instances.STATUS_VALUE_SCHEDULED,
+		Type:           db_instances.TYPE_CUSTOM,
+		CreatedAt:      time.Now().Format(time.RFC3339),
+		StartedAt:      "N/A",
+	})
+
+	if err != nil {
+		util.ThrowInternalServerErrorException(c, "Could not save item at the current time")
+		return
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:202]: Unnecessary return statement.

File: backend/pkg/instances/handler.go [hunk 8]
```diff
@@ -171,9 +240,85 @@ func DeleteContainerHandler(c *gin.Context) {
 	})
 
 	if err != nil {
-		util.ThrowServiceUnavailableException(c, "Could not schedule container at the current time")
+		util.ThrowServiceUnavailableException(c, "Could  schedule container deletion at the current time")
 		return
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:244]: Unnecessary return statement.

File: backend/pkg/instances/handler.go [hunk 9]
```diff
@@ -171,9 +240,85 @@ func DeleteContainerHandler(c *gin.Context) {
...
 	}
 
+	err = db_instances.DeleteInstanceById(userId, containerId)
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:247]: Decide which naming scheme the function parameter should have. UserID and UserSub is not very intuitive.

File: backend/pkg/instances/handler.go [hunk 10]
```diff
@@ -171,9 +240,85 @@ func DeleteContainerHandler(c *gin.Context) {
...
+//
+// @Router /instances [get]
+func GetUserInstances(c *gin.Context) {
+	userId := c.Request.Header.Get(constants.USER_ID_HEADER)
+
+	instances, err := db_instances.GetAllUserInstances(userId)
+
+	if err != nil {
+		util.ThrowInternalServerErrorException(c, "Could not fetch data")
+		return
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:277]: Unnecessary return statement.

File: backend/pkg/instances/handler.go [hunk 11]
```diff
@@ -171,9 +240,85 @@ func DeleteContainerHandler(c *gin.Context) {
...
+		util.ThrowBadRequestException(c, "No valid containerId passed")
+	}
+
+	instance, err := db_instances.GetInstanceById(userId, containerId)
+
+	if err != nil {
+		var notFoundErr *types.ResourceNotFoundException
+		if errors.As(err, &notFoundErr) {
+			util.ThrowNotFoundException(c, "Could not find item with given id")
+			return
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:317]: Unnecessary return statement.

File: backend/pkg/instances/handler.go [hunk 12]
```diff
@@ -171,9 +240,85 @@ func DeleteContainerHandler(c *gin.Context) {
...
+		}
+		util.ThrowInternalServerErrorException(c, "Could not delete entry at this time")
+		return
```
Comment (kontr0x) [inline backend/pkg/instances/handler.go:320]: Unnecessary return statement.

File: backend/pkg/webhook/handler.go [hunk 13]
```diff
@@ -55,20 +55,20 @@ func WebhookHandler(c *gin.Context) {
...
-    userBalance1, err := db_payment.GetUserBalance(userId)
-    if err != nil {
-        log.Error().Err(err).Msg("Failed to get user balance")
-        return
-    }
-    log.Info().Msgf("logUserBalance - User balance: %v", userBalance1)
+	userBalance1, err := db_payment.GetUserBalance(userId)
+	if err != nil {
+		log.Error().Err(err).Msg("Failed to get user balance")
+		return
```
Comment (kontr0x) [inline backend/pkg/webhook/handler.go:71]: Unnecessary return statement.

File: backend/pkg/webhook/handler.go [hunk 14]
```diff
@@ -55,20 +55,20 @@ func WebhookHandler(c *gin.Context) {
...
 }
 
 func logUserBalance(userId string) {
-    userBalance1, err := db_payment.GetUserBalance(userId)
-    if err != nil {
-        log.Error().Err(err).Msg("Failed to get user balance")
-        return
-    }
-    log.Info().Msgf("logUserBalance - User balance: %v", userBalance1)
+	userBalance1, err := db_payment.GetUserBalance(userId)
```
Comment (kontr0x) [inline backend/pkg/webhook/handler.go:68]: User balance 1? Why the numeric at the end.

**Review by kontr0x — CHANGES_REQUESTED** [review]
Left some comments. Please resolve and request review.

File: backend/pkg/db/tables/instances/instances.go (hunk 2 above)
Comment (coffeemakingtoaster) [reply backend/pkg/db/tables/instances/instances.go]: No

**Review by kontr0x — APPROVED** [review]
Looks good. PR can be merge!

**Pull request merged**